Added option to pull data from the Octopus Agile Outgoing API

tjsaunders opened the pull request:
Clearly open to suggestions, I would like to be able to change the block calculation to find the max rather than min for the Outgoing tariff block, but am interested in your input.

Thanks

tjsaunders committed:
Attempt 1 at adding Agile Outgoing

tjsaunders committed:
Attempt 2/fixes

Fix number of outputs and rename node for identifiability

tjsaunders committed:
nursery school fixes

semicolons and closed brackets

tjsaunders committed:
Update octopus.js

Ctrl+C the brackets, don't control X...

tjsaunders committed:
rename node

Renamed because no longer need identifiability

tjsaunders committed:
fix octopus.js

I fat finger removed the m from msg3

tjsaunders committed:
Trying to unbreak Agile

tjsaunders committed:
Update octopus.js

different Outgoing check because everything is coming back as agile

tjsaunders commented:
I should have added some evidence:
![image](https://user-images.githubusercontent.com/22834443/89343673-8c946880-d69c-11ea-976e-a572314e6c35.png)
This is my test flow, it has one node for Agile and one for Outgoing. Feeding into InfluxDB.
![image](https://user-images.githubusercontent.com/22834443/89343931-f6ad0d80-d69c-11ea-84e0-297264e09cff.png)
The Grafana output showing the 2 datasets in the InfluxDB with the 2 sources.

borpin commented on octopus.js:
```diff
@@ -105,17 +113,21 @@ module.exports = function(RED) {
-
-                                msg.payload.results.forEach(function(item, index) {
-                                    msg3.payload.push([{ value_inc_vat : item.value_inc_vat, 
+								if (APIurl.includes("OUTGOING")) {
```
Why not just use the `this.tariff == "OUTGOING"` comparison?

Even better is remove the if, and just set the value of `source` dependent on the Tariff selected (removes repeated code).

tjsaunders replied:
I tried this, but I think "this" doesn't refer to the node by this point in the function (and therefore this.tariff doesn't evaluate to anything)? The condition always evaluated to false. I couldn't think of a way to get the tariff value at that point in the code without just using the APIurl which I knew was there. I'm not massively familiar with js, so it may just be me not understanding.

tjsaunders replied:
This was the final 2 commits, the penultimate one was verifying what I thought was happening (both outputs swapped to Agile), and the final commit let them evaluate to different things.

borpin replied:
Looking at this (and to be honest my JS is a little minimal) I wonder why I ever moved it from `n.tariff` to `this.tariff`. `n.tariff` should be fine to use throughout. Try that.

tjsaunders replied:
> Looking at this (and to be honest my JS is a little minimal) I wonder why I ever moved it from `n.tariff` to `this.tariff`. `n.tariff` should be fine to use throughout. Try that.

Would you prefer I change this.region as well, or just this.tariff?

borpin replied:
That was my point. If you set a variable for `source` dependent on the tariff in use, you can use it without resorting to multiple 'if' statements and repeating code blocks.

borpin commented:
On looking for maximum, at line 103, there is an array of blocks, of length specified, that I then look for the `min`. It should be possible to look for a max block.  Probably needs a radio button in the UI to select max or min, then return the blocks on that basis.

Will need some rewording of the variables to make sense.

borpin commented:
I have also realised the `min_block_start` value should not be added to output1 as if there is more than one block it is just the last block. I'll raise an issue to remove it.

tjsaunders committed:
Try switching to n.tariff

borpin committed:
Merge branch 'master' into pr/10

tjsaunders committed:
min/max price block update

Documentation and code changes to add selection for min vs max block, and associated calcuation changes.

tjsaunders committed:
Merge branch 'dev' of https://github.com/tjsaunders/node-red-contrib-octopus into dev

tjsaunders committed:
Update documentation for clarity

tjsaunders committed:
Removed min_block_start from output 1

Not required

tjsaunders committed:
Consumption with API v0.1

tjsaunders committed:
Close all the brackets

tjsaunders committed:
Fix spelling mistakes

tjsaunders committed:
More spelling mistakes

tjsaunders committed:
Authentication on Consumption request

tjsaunders committed:
Tightened error handling

tjsaunders committed:
Removed overriding scope, added protocol to url

tjsaunders committed:
Trying to get errors to show

tjsaunders committed:
Same again

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Troubleshooting

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Debug logging

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Move node.send

tjsaunders committed:
Define msg4

tjsaunders committed:
Removed Debug logging

tjsaunders committed:
Changed .html to hide apikey

tjsaunders committed:
fix JSON Key spaces

tjsaunders committed:
Merge remote-tracking branch 'upstream/dev' into dev

tjsaunders committed:
Added Octopus Go (test)

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Update octopus.js

tjsaunders committed:
Remove random "n" from octopus.js

tjsaunders committed:
sysout apiurl

tjsaunders committed:
again

tjsaunders committed:
Added base url form field

tjsaunders committed:
Fix .html omission

tjsaunders committed:
Add identifier functionality

tjsaunders committed:
Update octopus.js

Fix misplaced identifier parameter

tjsaunders committed:
Update octopus.html

Fixed syntax

tjsaunders committed:
Update Influxdb output

tjsaunders committed:
Fix influxdb field syntax

tjsaunders committed:
Added datatype tag to influxdb output

tjsaunders committed:
Add tag for meter type. Change measurement name

tjsaunders committed:
Consumption Meter Type

tjsaunders committed:
type -> metertype

tjsaunders committed:
oop
